src/FluentValidation/Validators/NotEmptyValidator.cs

```diff
@@ -32,20 +33,28 @@ public class NotEmptyValidator : PropertyValidator, INotEmptyValidator {
 		}
 
 		protected override bool IsValid(PropertyValidatorContext context) {
-			switch (context.PropertyValue) {
+			return !IsEmpty(context.PropertyValue, _defaultValueForType);
+		}
+
+		internal static bool IsEmpty(object value, object defaultValueForType) {
+			// IStucturalComparable must come before ICollection.
+			// ImmutableArray<T> throws when checking its Count property if the array is uninitialized.
+			// Checking IStructuralComparable prevents this.
+			switch (value) {
 				case null:
 				case string s when string.IsNullOrWhiteSpace(s):
+				case IStructuralComparable sc when sc.CompareTo(defaultValueForType, Comparer.Default) == 0:
 				case ICollection c when c.Count == 0:
 				case Array a when a.Length == 0:
 				case IEnumerable e when !e.Cast<object>().Any():
-					return false;
+					return true;

```

Comment:
Have you considered using a switch expression with the `when` component moved to the action, instead of returning explicit booleans?

Something like:
```csharp
return value switch 
{
  null => true,
...
  Array a => a.Length == 0,
  IEnumerable e => !e.Cast<object>().Any(),
...
  _ => Equals(value, defaultValueForType),
}

Reply:
Let's keep the discussion focused on the problem at hand. Syntactic sugar can be addressed at a later stage.